data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3ObjectWorker.java

```diff
@@ -64,6 +67,81 @@ public S3ObjectWorker(final S3ObjectRequest s3ObjectRequest) {
+                               final S3ObjectReference s3ObjectReference,
+                               final BufferAccumulator<Record<Event>> bufferAccumulator,
+                               final SourceCoordinator<S3SourceProgressState> sourceCoordinator,
+                               final String partitionKey) throws IOException {
+        final S3InputFile inputFile = new S3InputFile(s3Client, s3ObjectReference, bucketOwnerProvider, s3ObjectPluginMetrics);
+        final String BUCKET = "bucket";
+        final String KEY = "key";
+        final String TIME = "time";
+        final String LENGTH = "length";
+        Map<String, Object> data = new HashMap<>();
+        data.put(BUCKET, s3ObjectReference.getBucketName());
+        data.put(KEY, s3ObjectReference.getKey());
+        data.put(TIME, inputFile.getLastModified());
+        data.put(LENGTH, inputFile.getLength());
+        Event event = JacksonEvent.builder()
+                .withEventType("event")
+                .withData(data)
+                .build();
+        final long s3ObjectSize = event.toJsonString().length();
+        if (acknowledgementSet != null) {
+            acknowledgementSet.add(event);
+        }
+        AtomicLong lastCheckpointTime = new AtomicLong(System.currentTimeMillis());
+        final AtomicInteger saveStateCounter = new AtomicInteger();
+        final Instant lastModifiedTime = inputFile.getLastModified();
+        final Instant now = Instant.now();
+        final Instant originationTime = (lastModifiedTime == null || lastModifiedTime.isAfter(now)) ? now : lastModifiedTime;
+        event.getMetadata().setExternalOriginationTime(originationTime);
+        event.getEventHandle().setExternalOriginationTime(originationTime);
+        try {
+            bufferAccumulator.add(new Record<>(event));
+        } catch (final Exception e) {
+            LOG.error("Failed writing S3 objects to buffer.", e);
+        }
+        if (acknowledgementSet != null && sourceCoordinator != null && partitionKey != null &&
+                (System.currentTimeMillis() - lastCheckpointTime.get() > DEFAULT_CHECKPOINT_INTERVAL_MILLS)) {
+            LOG.debug("Renew partition ownership for the object {}", partitionKey);
+            sourceCoordinator.saveProgressStateForPartition(partitionKey, null);
+            lastCheckpointTime.set(System.currentTimeMillis());
+            saveStateCounter.getAndIncrement();
+        }
+
+        try {
+            bufferAccumulator.flush();
+        } catch (final Exception e) {
+            LOG.error("Failed writing S3 objects to buffer.", e);
+        }
+
+        final int recordsWritten = bufferAccumulator.getTotalWritten();
+
+        if (recordsWritten == 0) {
+            LOG.warn("Failed to get metadata for S3 object: s3ObjectReference={}.", s3ObjectReference);
+            s3ObjectPluginMetrics.getS3ObjectNoRecordsFound().increment();
+        }
+        s3ObjectPluginMetrics.getS3ObjectSizeSummary().record(s3ObjectSize);
```

Comment:
I'm not sure it makes sense to include size metrics for metadata. We aren't processing the actual data.